### PR TITLE
CDAP-5437 Starting Zookeeper and Kafka clients in standalone mode when audit publishing is enabled

### PR DESCRIPTION
**Pull request by poornachandra**
JIRA - https://issues.cask.co/browse/CDAP-5437
Build - http://builds.cask.co/browse/CDAP-DUT3816-4

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java [hunk 1]
```diff
@@ -614,6 +615,11 @@ public boolean hasType(String datasetTypeName) {
   }
 
   private void publishAudit(Id.DatasetInstance datasetInstance, AuditType auditType) {
+    // Don't publish audit for system datasets admin operations, there can be a deadlock
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: but should we publish audit for system datasets at all?

Also, a little more information about the deadlock would be nice

File: cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java [hunk 2]
```diff
@@ -232,6 +258,7 @@ public void shutDown() {
         exploreExecutorService.stopAndWait();
       }
       exploreClient.close();
+      metadataService.stopAndWait();
```
Comment (bdmogal) [inline cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java]: why was this moved?

**Comment by bdmogal** [comment]
@poornachandra couple questions. Otherwise LGTM

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java (hunk 1 above)
Comment (russorat) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: IMO, system datasets should be published to the audit as well. If the dataset can be found via the tracker search, the audit logs should be published.

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: tracker should not show system datasets - am I missing something? In tracker, you're always in a namespace, and we don't surface system as a namespace in CDAP.

File: cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java (hunk 2 above)
Comment (poornachandra) [inline cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java]: Since metadata service depends on dataset service and transaction service, I moved the stop of metadata service before dataset service and transaction service were stopped.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java (hunk 1 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: Tracker is only one system that is used to viusalize audit messages right? Audit messages are for the whole CDAP system. I think it is okay to publish audit messages for system entities as long as it does not cause any issue. But since we don't publish audit in standalone mode now, we can revisit this. JIRA - https://issues.cask.co/browse/CDAP-5447

**Comment by poornachandra** [comment]
Build is green, merging.
